### ISSUE TITLE
Sensitive data masking is done in single thread with global mutex locked

### ISSUE THREAD

**Issue by tavplubix** [ClickHouse/ClickHouse#42776]
https://s3.amazonaws.com/clickhouse-test-reports/42749/e2252aabee5718647b73bf1f1ed312ec3137f1ff/stress_test__tsan_.html

```
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 4    0x000000001ab8147e in DB::CachedOnDiskReadBufferFromFile::nextImpl at ../src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp:773
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 4    0x000000001ab81f58 in DB::CachedOnDiskReadBufferFromFile::nextImplStep at ../src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp:853
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 2    0x000000000b973845 in DB::OwnSplitChannel::log at ../src/Loggers/OwnSplitChannel.cpp:43
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    0x000000000b973939 in DB::OwnSplitChannel::tryLogSplit at ../src/Loggers/OwnSplitChannel.cpp:53
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    0x000000000b973b38 in DB::OwnSplitChannel::logSplit at ../src/Loggers/OwnSplitChannel.cpp:87
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    0x000000000b96bc85 in non-virtual thunk to DB::OwnFormattingChannel::logExtended(DB::ExtendedLogMessage const&) ()
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    0x000000000b96bb1c in DB::OwnFormattingChannel::logExtended at ../src/Loggers/OwnFormattingChannel.cpp:16
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    Poco::FileChannel::log at ../contrib/poco/Foundation/src/FileChannel.cpp:107
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    Poco::ScopedLock<Poco::FastMutex>::ScopedLock at ../contrib/poco/Foundation/include/Poco/ScopedLock.h:38
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    Poco::FastMutex::lock at ../contrib/poco/Foundation/include/Poco/Mutex.h:238
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    0x000000002038f87e in Poco::MutexImpl::lockImpl at ../contrib/poco/Foundation/include/Poco/Mutex_POSIX.h:59
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    0x000000000b642a73 in pthread_mutex_lock ()
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    0x00007f9dbdb6e0a3 in pthread_mutex_lock () from /lib/x86_64-linux-gnu/libpthread.so.0
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    0x00007f9dbdb76170 in __lll_lock_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 2    0x000000000b9737b9 in DB::OwnSplitChannel::log at ../src/Loggers/OwnSplitChannel.cpp:34
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    DB::SensitiveDataMasker::wipeSensitiveData at ../src/Common/SensitiveDataMasker.cpp:167
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    0x000000000b7ddb9e in DB::SensitiveDataMasker::MaskingRule::apply at ../src/Common/SensitiveDataMasker.cpp:66
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    0x0000000020d3ced5 in re2::RE2::GlobalReplace at ../contrib/re2/re2/re2.cc:449
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    0x0000000020d3be70 in re2::RE2::Match at ../contrib/re2/re2/re2.cc:744
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    0x0000000020d5ef86 in re2::Prog::SearchDFA at ../contrib/re2/re2/dfa.cc:1868
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    re2::DFA::Search at ../contrib/re2/re2/dfa.cc:1770
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    0x0000000020d5eb23 in re2::DFA::FastSearchLoop at ../contrib/re2/re2/dfa.cc:1591
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    0x0000000020d5e341 in re2::DFA::SearchTFT at ../contrib/re2/re2/dfa.cc:1563
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    0x0000000020d65b92 in re2::DFA::InlinedSearchLoop<true, false, true> at ../contrib/re2/re2/dfa.cc:1445
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 1    0x000000000b690151 in __tsan_write8 ()
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 4    0x000000001ab84229 in DB::CachedOnDiskReadBufferFromFile::nextImplStep at ../src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp:1002
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 2    0x000000000b973845 in DB::OwnSplitChannel::log at ../src/Loggers/OwnSplitChannel.cpp:43
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 2    0x000000000b973939 in DB::OwnSplitChannel::tryLogSplit at ../src/Loggers/OwnSplitChannel.cpp:53
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 2    0x000000000b973b38 in DB::OwnSplitChannel::logSplit at ../src/Loggers/OwnSplitChannel.cpp:87
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 2    0x000000000b96bc85 in non-virtual thunk to DB::OwnFormattingChannel::logExtended(DB::ExtendedLogMessage const&) ()
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 2    0x000000000b96bb1c in DB::OwnFormattingChannel::logExtended at ../src/Loggers/OwnFormattingChannel.cpp:16
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 2    Poco::FileChannel::log at ../contrib/poco/Foundation/src/FileChannel.cpp:107
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 2    Poco::ScopedLock<Poco::FastMutex>::ScopedLock at ../contrib/poco/Foundation/include/Poco/ScopedLock.h:38
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 2    Poco::FastMutex::lock at ../contrib/poco/Foundation/include/Poco/Mutex.h:238
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 2    0x000000002038f87e in Poco::MutexImpl::lockImpl at ../contrib/poco/Foundation/include/Poco/Mutex_POSIX.h:59
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 2    0x000000000b642a73 in pthread_mutex_lock ()
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 2    0x00007f9dbdb6e0a3 in pthread_mutex_lock () from /lib/x86_64-linux-gnu/libpthread.so.0
| | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | 2    0x00007f9dbdb76170 in __lll_lock_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
```

**Comment by alexey-milovidov** [comment]
It also does not work at all: https://github.com/ClickHouse/ClickHouse/issues/41418
CC @filimonov

**Comment by filimonov** [comment]
No sure if i understand the problem here... SensitiveDataMasker does not have any locks and normally should be very fast (much faster than i/o).

**Comment by KochetovNicolai** [comment]
There is also a race condition found 
https://s3.amazonaws.com/clickhouse-test-reports/57855/a4d287b5767efb0c2587bd46c03d3f6600929515/stress_test__tsan_/stderr.log

**Comment by alexey-milovidov** [comment]
1. From the original report, it is clear that the mutex belongs to `FileSplitChannel`. There is no mutex in `OwnSplitChannel`. Maybe there is a mutex somewhere above, but I doubt it. So, the original issue is no longer correct.

2.
> It also does not work at all

This problem was fixed.

3.
> and normally should be very fast (much faster than i/o).

There is no such thing as "much faster than I/O" (it is a common cliche from old historical times).
Moreover, the SensitiveDataMasker can contain an arbitrary number of regular expressions.
Even a single regular expression can be slower than a single SSD speed,
but even this is irrelevant because - who said that logging does any I/O?

4. There is also a race condition found

It was found recently, but we reverted the offending PR almost instantly.
